### PR TITLE
Configurable cluster blacklist TTL

### PR DESCRIPTION
**Pull request by BCathcart**
Allows cluster admins to configure the blacklist TTL as needed to allow sufficient time for `CLUSTER FORGET` to be executed on every node in the cluster.

Config name `cluster-blacklist-ttl`; unit seconds; deault 60.

### CODE REVIEW

File: src/cluster_legacy.h
Comment (BCathcart) [inline src/cluster_legacy.h]: I doesn't seem like having a default config value defined in a relevant header file is a common practice, so I can remove this if that's preferred

Comment (hpatro) [reply src/cluster_legacy.h]: It's easier to see the number directly where the config is defined. So, I would say remove this.

File: valkey.conf
Comment (hpatro) [inline valkey.conf]: Could you put the default value here, even if it's commented.

**Comment by hpatro** [comment]
Also you would need to signoff https://github.com/valkey-io/valkey/runs/26959011274

**Comment by hpatro** [comment]
@zuiderkwast We discussed this briefly during the summit. PTAL.

**Comment by zuiderkwast** [comment]
Hi @BCathcart! Have you experienced this problem after we added blacklist propagation in the cluster bus?

**Comment by BCathcart** [comment]
> Hi @BCathcart! Have you experienced this problem after we added blacklist propagation in the cluster bus?

Hey @zuiderkwast. Not exactly. We piggy-backed some additional functionality internally on top of the blacklist, which was the real motivation for making it configurable. It sounds like it would be a nice-to-have in OSS as well but maybe the blacklist propagation addresses this enough already.

**Comment by zuiderkwast** [comment]
> We piggy-backed some additional functionality internally on top of the blacklist

Interesting. What was that?

If it's not a problem anyone is experiencing, I don't think we need to solve it. :)

In general, I think we should avoid adding configs. If this is actually a problem in a large cluster, could we let the value depend on parameters such as the size of the cluster and the node timeout instead? It's possible to estimate how long it takes for the blacklist to be propagated to all nodes in a cluster.

**Comment by BCathcart** [comment]
> Interesting. What was that?

It was a quick fix to help address a cluster message formatting mismatch between versions. It's very irrelevant to OSS so I won't get into the details.

I'm not feeling a strong need to make this configurable either now that I know about the blacklist propagation. Scaling the ttl based on cluster size and existing parameters seems reasonable to me, but I have no evidence the current 60 seconds is ever not enough time for the blacklist propagation in practice, so I don't think we need to jump into anything for now.

@hpatro @madolson Any more thoughts on this?

**Comment by hpatro** [comment]
@zuiderkwast In case of partition, the propagation might not arrive within a hard coded time period. I think making the parameter flexible for operator makes life easier.

**Review by enjoy-binbin — COMMENTED** [review]
For the record, we had the same configuration item in ours fork, so i think it is ok in my head.

In some cases, the node may not forget the other node in time (like partition or admin tools bug), and we don't it to add it back (at a configurable time).

**Comment by madolson** [comment]
> If it's not a problem anyone is experiencing, I don't think we need to solve it. :)

Oh, we've had this problem long before we were doing the piggy-backing :). I think it's especially bad in large clusters. Our internal control plane folks have asked for this feature for awhile.

> It's possible to estimate how long it takes for the blacklist to be propagated to all nodes in a cluster.

This would be more ideal, but I'm not sure we'll ever be able to get it right. I suppose my opinion about configs is a bit more nuanced, which is that I am fine with lots of configs that most people will never need to touch, but are nice to have in case they need to touch them. I would also prefer to limit the number of configs on the datapath, but I really don't mind administration configs as much.

**Comment by zuiderkwast** [comment]
I'm fine with adding the config. I just wanted to know that it's useful before we add it.

Just to be clear: If a user changes the config to one hour, does CLUSTER FORGET and afterwards changes the config back to one minute, then the timestamp in the blacklist will be one hour. The user can't add the node again during this time.

File: src/config.c
Comment (zuiderkwast) [inline src/config.c:3214]: This change from LONG_MAX to ULONG_MAX changes the valid range for all these configs. Is it worth mentioning in the release notes?

The hypothetical incompatibility I can imagine: Someone configures one of these to LONG_MAX + 1, then downgrades to 7.2.5, then the config will be invalid...?

**Review by zuiderkwast — APPROVED** [review]
Implementation LGTM

File: src/config.c
Comment (madolson) [reply src/config.c:3214]: I don't think we should be changing these configs. They were proposed in another PR as well, and done just for consistency with the max value. These values are no really reasonable at U_long max anyways.

Comment (BCathcart) [reply src/config.c:3214]: Updated to leave existing configs alone like was done on the other PR

File: valkey.conf
Comment (madolson) [inline valkey.conf]:
```suggestion
# The cluster blacklist is used when removing a node from the cluster completely. 
# When CLUSTER FORGET is called for a node, that node is put into the blacklist for
# some time so that when gossip messages are received from other nodes that still
# remember it, it is not re-added. This gives time for CLUSTER FORGET to be sent to
# every node in the cluster. The blacklist TTL is 60 seconds by default, which should
# sufficient for most clusters, but you may considering increasing this if you see
# nodes getting re-added while using CLUSTER FORGET.
```

Just make it a bit more prescriptive.

**Review by madolson — APPROVED** [review]
LGTM, just added some wording updates.

**Comment by madolson** [comment]
@valkey-io/core-team Not for valkey 8.0, but please vote if we want to add this config to future versions.

**Review by hpatro — APPROVED** [review]
Thanks @BCathcart 👍

**Comment by hpatro** [comment]
@madolson we won't even add config changes to 8.0 further?

**Comment by BCathcart** [comment]
The commits from my quick edits on github weren't automatically signed off so I had to do a little rebase. Hopefully it's good to go now.

File: src/config.c
Comment (enjoy-binbin) [inline src/config.c]: this link was merged in another PR, care to do a rebase (or a merge)?

**Review by enjoy-binbin — APPROVED** [review]
the code LGTM.

**Comment by zuiderkwast** [comment]
> @valkey-io/core-team Not for valkey 8.0, but please vote if we want to add this config to future versions.

It seems ready to merge. I can see 5 core team members approving or upvoting. If we merge it before we cut the release branch, I see no reason why it can't be included in 8.0.

**Comment by madolson** [comment]
Yeah, I guess it's before the cutoff so we can add it.

**Pull request merged**
